File: backend/src/main/java/com/uth/confms/storage/service/BackupService.java
package com.uth.confms.storage.service;

import com.uth.confms.storage.entity.BackupRecord;

import java.util.List;

/**
 * Service để quản lý backup và recovery cho file storage
 */
public interface BackupService {

    /**
     * Tạo backup cho một conference cụ thể
     *
     * @param conferenceId ID của conference
     * @return BackupRecord chứa thông tin backup
     */
    // Tạo bản sao lưu cho conference
    BackupRecord backupConference(Long conferenceId);

    /**
     * Tạo full backup cho tất cả files
     *
     * @return BackupRecord chứa thông tin backup
     */
    // Tạo bản sao lưu toàn bộ
    BackupRecord backupAll();

    /**
     * Restore files từ backup
     *
     * @param backupId ID của backup record
     * @return true nếu restore thành công
     */
    // Khôi phục từ bản sao lưu
    boolean restoreBackup(Long backupId);

    /**
     * Restore files cho một conference từ backup gần nhất
     *
     * @param conferenceId ID của conference
     * @return true nếu restore thành công
     */
    // Khôi phục dữ liệu conference
    boolean restoreConference(Long conferenceId);

    /**
     * Lấy danh sách backups cho một conference
     *
     * @param conferenceId ID của conference
     * @return List of BackupRecord
     */
    // Lấy lịch sử sao lưu
    List<BackupRecord> getBackupHistory(Long conferenceId);

    /**
     * Lấy danh sách tất cả backups
     *
     * @return List of BackupRecord
     */
    // Lấy danh sách tất cả bản sao lưu
    List<BackupRecord> getAllBackups();

    /**
     * Xóa backup cũ (cleanup)
     *
     * @param olderThanDays Xóa backups cũ hơn số ngày này
     * @return Số lượng backups đã xóa
     */
    // Dọn dẹp các bản sao lưu cũ
    int cleanupOldBackups(int olderThanDays);
}

File: frontend/src/services/backup.service.ts
import apiClient from './api'

export interface BackupRecord {
    id: number
    conferenceId?: number
    backupType: 'DATABASE' | 'STORAGE' | 'FULL'
    filePath: string
    fileSize: number
    status: 'PENDING' | 'COMPLETED' | 'FAILED'
    createdAt: string
    createdBy: number
}

/**
 * Backup Service - Admin tool for data protection (API v1)
 */
export const backupService = {
    /**
     * Lấy danh sách tất cả các bản sao lưu
     * GET /api/storage/backup
     */
    getAllBackups: async (): Promise<BackupRecord[]> => {
        const response = await apiClient.get<any>('/storage/backup')
        return response.data?.data || response.data || []
    },

    /**
     * Lấy lịch sử sao lưu của một hội nghị cụ thể
     * GET /api/storage/backup/conference/{id}
     */
    getConferenceBackups: async (conferenceId: number): Promise<BackupRecord[]> => {
        const response = await apiClient.get<any>(`/storage/backup/conference/${conferenceId}`)
        return response.data?.data || response.data || []
    },

    /**
     * Tạo bản sao lưu toàn bộ hệ thống
     * POST /api/storage/backup/all
     */
    backupAll: async (): Promise<BackupRecord> => {
        const response = await apiClient.post<any>('/storage/backup/all')
        return response.data?.data || response.data
    },

    /**
     * Tạo bản sao lưu cho một hội nghị cụ thể
     * POST /api/storage/backup/conference/{id}
     */
    backupConference: async (conferenceId: number): Promise<BackupRecord> => {
        const response = await apiClient.post<any>(`/storage/backup/conference/${conferenceId}`)
        return response.data?.data || response.data
    },

    /**
     * Khôi phục từ một bản sao lưu cụ thể
     * POST /api/storage/backup/{id}/restore
     */
    restore: async (backupId: number): Promise<boolean> => {
        const response = await apiClient.post<any>(`/storage/backup/${backupId}/restore`)
        return response.data?.data ?? response.data ?? false
    },

    /**
     * Khôi phục hội nghị từ bản sao lưu mới nhất
     * POST /api/storage/backup/conference/{id}/restore
     */
    restoreConference: async (conferenceId: number): Promise<boolean> => {
        const response = await apiClient.post<any>(`/storage/backup/conference/${conferenceId}/restore`)
        return response.data?.data ?? response.data ?? false
    },
}
